### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by Boanta-V-Razvan**

### CODE REVIEW

File: app/auth/auth.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    if not user:
+        return None
+    if not bcrypt.verify(password, user.password):
+        return None
+    return user
+
+
+async def check_token(token: str = Depends(scheme)):
+    try:
+        payload = jwt.decode(token, jwt_secret, algorithms=['HS256'])
```
Comment (BogdanChivescu) [inline app/auth/auth.py:24]: no expiration in token?

File: app/database/database.py [hunk 2]
```diff
@@ -22,3 +25,17 @@ def get_item(self, item_id, model):
         data = self.db[model.__name__].find_one({"_id": ObjectId(item_id)})
         data = parse_obj_as(model, data)
         return data
+
+    def get_user(self, username, _id=None):
+        look_dict = {"username": username}
+        if _id:
+            look_dict = {"$and": [{"username": username}, {"_id": ObjectId(_id)}]}
```
Comment (BogdanChivescu) [inline app/database/database.py:32]: you don't need the username anymore if you have the id

File: app/database/database.py [hunk 3]
```diff
@@ -22,3 +25,17 @@ def get_item(self, item_id, model):
...
+        user = self.db.users.find_one(look_dict)
+        if user:
+            user = User(
```
Comment (BogdanChivescu) [inline app/database/database.py:35]: use `parse_obj_as` from pydantic

File: app/routers/audit.py [hunk 4]
```diff
@@ -14,25 +14,28 @@
...
     @router.get("/audits/{audit_id}")
-    def read_audit(self, audit_id: str):
+    def read_audit(self, audit_id: str, user: User = Depends(auth.check_token)):
         try:
-            return self.storage.get_item(audit_id, Audit)
+            if user:
+                return self.storage.get_item(audit_id, Audit)
         except Exception as err:
             logging.error(err)
             raise HTTPException(status_code=404, detail="Item not found")
```
Comment (BogdanChivescu) [inline app/routers/audit.py:30]: we should return another response if the user is not logged in. Applies everywhere

File: app/routers/audit.py [hunk 5]
```diff
@@ -14,25 +14,28 @@
...
-        return self.storage.list_items(Audit)
+    def read_all_audits(self,  user: User = Depends(auth.check_token)):
+        if user:
+            return self.storage.list_items(Audit)
 
     @router.post("/audits")
-    def create_audit(self, audit: Audit):
-        audit_id = self.storage.create_item(audit)
-        return audit_id
+    def create_audit(self, audit: Audit, user: User = Depends(auth.check_token)):
```
Comment (BogdanChivescu) [inline app/routers/audit.py:38]: isn't there a way to apply `auth.check_token` to the whole router?

File: app/routers/user.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+class UserController:
+    def __init__(self):
+        self.storage = Storage()
+
+    @router.post('/user/token')
+    async def generate_token(self, form_data: OAuth2PasswordRequestForm = Depends()):
+        user = await auth.authenticate_user(form_data.username, form_data.password)
+
+        if not user:
+            return {'Error': 'Could not authenticate user, wrong credentials'}
```
Comment (BogdanChivescu) [inline app/routers/user.py:25]: raise Exception?

File: app/routers/user.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+router = InferringRouter()
+
+
+@cbv(router)
+class UserController:
+    def __init__(self):
+        self.storage = Storage()
+
+    @router.post('/user/token')
```
Comment (BogdanChivescu) [inline app/routers/user.py:20]: `/login`?

File: app/routers/user.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        if not user:
+            return {'Error': 'Could not authenticate user, wrong credentials'}
+        payload = dict(user)
+        payload.pop("password",None)
+        token = jwt.encode(payload, auth.jwt_secret)
+        return {'access_token': token, 'token_type': 'bearer'}
+
+
+    @router.post('/user')
+    async def create_user(self, user: User):
```
Comment (BogdanChivescu) [inline app/routers/user.py:33]: comment this out for now, we'll create users manually

File: app/routers/audit.py (hunk 5 above)
Comment (BogdanChivescu) [reply app/routers/audit.py:38]: or the whole application for that matter, apart from login

File: app/auth/auth.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        user = storage.get_user(
+            username=payload['username'],
+            _id=payload['id']
+        )
+    except:
```
Comment (BogdanChivescu) [inline app/auth/auth.py:29]: what will raise an exception here? `jwt.decode`?
`storage.get_user` can return None

**Review by BogdanChivescu — CHANGES_REQUESTED** [review]

File: app/routers/user.py (hunk 8 above)
Comment (Boanta-V-Razvan) [reply app/routers/user.py:33]: Unless you add a hashed password it will not work

File: app/routers/audit.py (hunk 5 above)
Comment (Boanta-V-Razvan) [reply app/routers/audit.py:38]: recommended for each

File: app/routers/user.py (hunk 7 above)
Comment (Boanta-V-Razvan) [reply app/routers/user.py:20]: That's more a front end path stuff , here you just give a token of access you dont create a session or something like that

File: app/auth/auth.py (hunk 1 above)
Comment (Boanta-V-Razvan) [reply app/auth/auth.py:24]: Is it really necessary? Needs more discussion

File: app/database/database.py (hunk 2 above)
Comment (BogdanChivescu) [reply app/database/database.py:32]: I meant in `look_dict`

File: app/routers/audit.py (hunk 5 above)
Comment (BogdanChivescu) [reply app/routers/audit.py]: https://fastapi.tiangolo.com/tutorial/bigger-applications/#another-module-with-apirouter

File: app/auth/auth.py (hunk 1 above)
Comment (BogdanChivescu) [reply app/auth/auth.py:24]: we can leave it without expiry for now, but I still think we;ll need to add it later

**Pull request merged**